### PR TITLE
Update position before ReadAsync starts, but fix it after incomplete read

### PR DESCRIPTION
**Pull request by adamsitnik**

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/IO/Strategies/OSFileStreamStrategy.cs [hunk 1]
```diff
@@ -99,6 +99,10 @@ public unsafe sealed override long Length
             }
         }
 
+        // in case of concurrent incomplete reads, there can be multiple threads trying to update the position
+        // at the same time. That is why we are using Interlocked here.
+        internal void OnIncompleteRead(int expectedBytesRead, int actualBytesRead) => Interlocked.Add(ref _filePosition, actualBytesRead - expectedBytesRead);
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/IO/Strategies/OSFileStreamStrategy.cs:104]: @stephentoub I've used `Interlocked` here as the new `IncompleteReadCantSetPositionBeyondEndOfFile` test was simply failing without it. I know that we are never going to be 100% thread safe, but the alternative was to call `Length` which would result in an extra sys-call.

```cs
_filePosition = Math.Min(Length, _filePosition - expectedBytesRead + actualBytesRead);
```

**Comment by adamsitnik** [comment]
the test is failing on Windows 7 and 8, I am going to apply no merge until I fix it

**Comment by adamsitnik** [comment]
Setting file position and length via `UpdateLengthOnChangePosition` before starting the async write operation could lead to potentially invalid `Length` value in cases when `WriteAsync` has failed (with an error or was cancelled)

I've disabled `Length` caching for files open for writing. We no longer need `Length` in `ReadAsync` so we are not regressing any important scenario.

**Comment by adamsitnik** [comment]
@stephentoub all tests are passing, PTAL

File: src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs [hunk 2]
```diff
@@ -286,6 +288,13 @@ internal static unsafe (SafeFileHandle.OverlappedValueTaskSource? vts, int error
                 vts.Dispose();
                 throw;
             }
+            finally
+            {
+                if (errorCode != Interop.Errors.ERROR_IO_PENDING && errorCode != Interop.Errors.ERROR_SUCCESS)
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs:293]: Nit: this function is initializing errorCode to '0' but then checking against 'ERROR_SUCCESS'.  Style-wise, it'd be nice if they were consistent.

File: src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs [hunk 3]
```diff
@@ -69,35 +58,22 @@ public override Task WriteAsync(byte[] buffer, int offset, int count, Cancellati
...
 
         private Exception HandleIOError(long positionBefore, int errorCode)
         {
-            if (!_fileHandle.IsClosed && CanSeek)
+            if (_fileHandle.CanSeek)
             {
                 // Update Position... it could be anywhere.
-                _filePosition = positionBefore;
+                Interlocked.Exchange(ref _filePosition, positionBefore);
             }
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs:77]: This is precarious regardless of the interlocked, e.g. if someone did kick off concurrent operations,  they could all fight to reset this to their starting position. It's not clear there's a "right" answer.

**Review by stephentoub — APPROVED** [review]

**Comment by gewarren** [comment]
Did this PR effectively negate the breaking change listed here? https://github.com/dotnet/docs/issues/23858

**Comment by adamsitnik** [comment]
> Did this PR effectively negate the breaking change listed here? https://github.com/dotnet/docs/issues/23858

Yes and no.

IIRC we wanted to update the position after the operations are completed, but then we realized that some users use `FileStream` for parallel reads and writes despite the fact it's not supported (`FileStream` is officially not thread safe). We did not want to break these users, so we've decided to optimistically set the position when the operation is being started. We just assume that whole buffer will be read/written, but when the operation finishes and it turns out to not be true, we update the position accordingly.

My main goal with https://github.com/dotnet/docs/issues/23858 was to give the users a clear message: stop relying on that, don't use `FileStream` for parallel file operations (we have provided a new API designed for that: `RandomAccess`).

**Pull request merged**